### PR TITLE
feat!: major v3 refactor

### PR DESCRIPTION
**Pull request by stipsan**
This refactor narrows down the API surface to what we use in `sanity`, fixes issues with Strict Mode in React 19, improves perf in React 18 by calling `store.subscription.unsubscribe` when the `useSyncExternalStore` hook calls `subscribe`, uses the native `useSyncExternalStore` hook, and more!

A test branch with this version is available here: https://github.com/sanity-io/sanity/pull/6883/files?diff=unified&w=1

It's probably best to rebase + merge this PR so the changelog is more detailed.
There's a lot of information in the body of some of the commits btw.

### CODE REVIEW

**Comment by stipsan** [comment]
@SocketSecurity ignore npm/esbuild@0.21.4

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-rx",
-  "version": "2.1.3",
+  "version": "3.0.0-crx-749.1",
   "description": "React + RxJS = <3",
   "keywords": [
```
Comment (bjoerge) [inline package.json:5]: Can probably remove `use-sync-external-store` from the keywords here

File: src/useObservable.ts [hunk 2]
```diff
@@ -1,98 +1,69 @@
...
-  deps: DependencyList,
-  initialValue?: T | (() => T),
-) {
-  return useObservable(
-    useMemo(() => getValue(observableOrFactory), deps),
-    initialValue,
-  )
-}
+/** @internal */
+export type UnboxObservable<T> = T extends Observable<infer U> ? U : never
```
Comment (bjoerge) [inline src/useObservable.ts]: There's an [`ObservedValueOf`](https://rxjs.dev/api/index/type-alias/ObservedValueOf) type exported from rxjs that captures a lot more value types than just observables.

It probably makes sense for us to keep it narrow for now though, but at least adding a separate condition for `InteropObservable` here can potentially save us some compatibility issues later on.

File: src/useObservableEvent.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    return () => subscription.unsubscribe()
+  }, [callback, calls$])
+
+  return call
+}
+
+/**
+ * This is a ponyfill of the upcoming `useEffectEvent` hook that'll arrive in React 19
+ */
+function useEffectEvent<
```
Comment (bjoerge) [inline src/useObservableEvent.ts:26]: Nice

**Review by bjoerge — COMMENTED** [review]
This is great, thank you so much for doing this <3

A few thoughts:
- We should fix the website. It's running against the latest 2.x and won't reflect what this package is actually about and the APIs it exports. We can probably scale it down to a bare minimum (or replace it with a better readme). Not a blocker for release though I'd say.
- I might re-introduce the `reactiveComponent` (or - at least look into whether it makes sense to re-introduce it) at some point later as I'm a big fan of this pattern (and use it quite extensively in a few of my own projects). But understand why it was removed.
- Should we move to vitest? (yes, I'll make a PR)

**Comment by stipsan** [comment]
> * We should fix the website. It's running against the latest 2.x and won't reflect what this package is actually about and the APIs it exports. We can probably scale it down to a bare minimum (or replace it with a better readme). Not a blocker for release though I'd say.

Sure, I'll do that now 👍  I've added the Vercel deployment to the list over required checks to ensure I don't forget :D 

> * I might re-introduce the `reactiveComponent` (or - at least look into whether it makes sense to re-introduce it) at some point later as I'm a big fan of this pattern (and use it quite extensively in a few of my own projects). But understand why it was removed.

Yeah, sure! I would do it for a react 19 release. In react 19 you don't need `React.forwardRef` wrappers, and you have a bunch of new APIs for async logic, like `startTransition(async () => {await fetch(url, {method: 'POST'})})`, `useOptimisticState`, `React.use` and more. Having a new `reactiveComponent` pattern that let you take advantage of these in rxjs in an elegant way would be very exciting ✨ 

> * Should we move to vitest? (yes, I'll make a PR)

Brilliant 💖

File: package.json (hunk 1 above)
Comment (stipsan) [reply package.json:5]: Done!

File: src/useObservable.ts (hunk 2 above)
Comment (stipsan) [reply src/useObservable.ts]: I refactored to `ObservedValueOf` and it worked. I couldn't get `InteropObservable` to work well as it caused type errors in the setup that adds `shareReplay` and such 🤔

**Comment by stipsan** [comment]
@SocketSecurity ignore npm/fsevents@1.2.13

**Review by bjoerge — APPROVED** [review]

**Pull request merged**